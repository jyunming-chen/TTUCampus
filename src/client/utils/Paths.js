// @ts-check

import Path from './Path';

/**
 * @typedef {Object} ConfigPath
 * @prop {string} from
 * @prop {string} to
 * @prop {number[][]} points
 */

export default class Paths {
  /**
   * @param {ConfigPath[]} paths
   */
  constructor(paths) {
    /**
     * @type {Record<string, Record<string, Path>>}
     */
    this.map = {};
    paths.forEach(path => {
      this.setPath(path.from, path.to, path.points);
      this.setPath(path.to, path.from, path.points.slice().reverse());
    });
  }

  /**
   * @param {string} from
   * @param {string} to
   * @param {number[][]} points
   */
  setPath(from, to, points) {
    if (!this.map[from]) {
      this.map[from] = {};
    }
    this.map[from][to] = new Path(points);
  }

  /**
   * @param {string} from
   * @param {string} to
   */
  getPath(from, to) {
    if (!this.map[from]) {
      return undefined;
    }
    return this.map[from][to];
  }
}
