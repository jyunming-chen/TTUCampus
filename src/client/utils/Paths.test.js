import Paths from './Paths';

const paths = new Paths([
  {
    from: 'A',
    to: 'B',
    points: [[0, 0], [1, 1], [2, 2]],
  },
  {
    from: 'A',
    to: 'C',
    points: [[0, 0], [-1, -2], [-3, -4]],
  },
]);

test('new Paths(configPaths)', () => {
  expect(paths).toMatchSnapshot();
});

test('#getPath()', () => {
  expect(paths.getPath('unknown', 'unknown')).toBeUndefined();
  expect(paths.getPath('A', 'B')).toMatchSnapshot();
});
