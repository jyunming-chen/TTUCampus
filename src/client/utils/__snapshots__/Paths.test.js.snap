// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#getPath() 1`] = `
Path {
  "distance": 2.8284271247461903,
  "distances": Array [
    0,
    1.4142135623730951,
    2.8284271247461903,
  ],
  "points": Array [
    Vector3 {
      "x": 0,
      "y": 0,
      "z": 0,
    },
    Vector3 {
      "x": 1,
      "y": 0,
      "z": -1,
    },
    Vector3 {
      "x": 2,
      "y": 0,
      "z": -2,
    },
  ],
}
`;

exports[`new Paths(configPaths) 1`] = `
Paths {
  "map": Object {
    "A": Object {
      "B": Path {
        "distance": 2.8284271247461903,
        "distances": Array [
          0,
          1.4142135623730951,
          2.8284271247461903,
        ],
        "points": Array [
          Vector3 {
            "x": 0,
            "y": 0,
            "z": 0,
          },
          Vector3 {
            "x": 1,
            "y": 0,
            "z": -1,
          },
          Vector3 {
            "x": 2,
            "y": 0,
            "z": -2,
          },
        ],
      },
      "C": Path {
        "distance": 5.06449510224598,
        "distances": Array [
          0,
          2.23606797749979,
          5.06449510224598,
        ],
        "points": Array [
          Vector3 {
            "x": 0,
            "y": 0,
            "z": 0,
          },
          Vector3 {
            "x": -1,
            "y": 0,
            "z": 2,
          },
          Vector3 {
            "x": -3,
            "y": 0,
            "z": 4,
          },
        ],
      },
    },
    "B": Object {
      "A": Path {
        "distance": 2.8284271247461903,
        "distances": Array [
          0,
          1.4142135623730951,
          2.8284271247461903,
        ],
        "points": Array [
          Vector3 {
            "x": 2,
            "y": 0,
            "z": -2,
          },
          Vector3 {
            "x": 1,
            "y": 0,
            "z": -1,
          },
          Vector3 {
            "x": 0,
            "y": 0,
            "z": 0,
          },
        ],
      },
    },
    "C": Object {
      "A": Path {
        "distance": 5.06449510224598,
        "distances": Array [
          0,
          2.8284271247461903,
          5.06449510224598,
        ],
        "points": Array [
          Vector3 {
            "x": -3,
            "y": 0,
            "z": 4,
          },
          Vector3 {
            "x": -1,
            "y": 0,
            "z": 2,
          },
          Vector3 {
            "x": 0,
            "y": 0,
            "z": 0,
          },
        ],
      },
    },
  },
}
`;
